### DISCUSSION TITLE
[Note Template]

### DISCUSSION THREAD

**Discussion by Dcrizz** [windingwind/zotero-better-notes#619] [Note Templates]
## Template Type

MDFileheader 

## Desired Template Performance

Insert all details I want into notes metadata upon import to obsidian

### Describe what you want.

I have tried the authors and creators fields but it always comes up black upon import and I am not sure how to do this with the  **noteItem.parentItem? noteItem.parentItem.getField** format in the MDFileheader. The fields I have in my current template work, but the additional fields I have listed below I can seem to get to work. I have almost no knowledge of coding or this syntax and have tried numerous ways via trial and error and got some to work but the fields below are beyond my knowledge. Any help would be greatly appreciated 

Additionally can you explain to me how you use this in that was in the templates guidance section

**Find available fields of the selected item with the code below:**

const item = ZoteroPane.getSelectedItems().shift();
const usedFields = item.getUsedFields();
Zotero.ItemFields.getAll()
  .filter((e) => usedFields.indexOf(e.id) >= 0)
  .map((e) => e.name);
The result is like this (depending on the item you select):

[
    "0": "title"
    "1": "date"
    "2": "language"
    "3": "shortTitle"
    "4": "libraryCatalog"
    "5": "url"
    "6": "accessDate"
    "7": "pages"
    "8": "conferenceName"
]

**Fields I want to add:**

#### Authors (can't seem to get this to work)

<p>
  $
  {topItem
    .getCreators()
    .map((v) => v.firstName + " " + v.lastName)
    .join("; ")}
</p>

####The format for conditionals such as:

##### Journal
<p>
  $
  {(() => {
    if (topItem.itemType === "conferencePaper") {
      return;
      topItem.getField("conferenceName") ||
        topItem.getField("proceedingsTitle");
    }
    if (topItem.itemType === "journalArticle")
      return topItem.getField("publicationTitle");
    if (topItem.itemType === "report") return topItem.getField("institution");
    return topItem.getField("publicationTitle");
  })()}
</p>

 ##### Pdf link
<p>
  $
  {((topItem) => {
    const getPDFLink = (_item) => {
      let libraryID = _item.libraryID;
      let library = Zotero.Libraries.get(libraryID);
      let itemKey = _item.key;
      let itemLink = "";
      if (library.libraryType === "user") {
        itemLink = `zotero://open-pdf/library/items/${itemKey}`;
      } else if (library.libraryType === "group") {
        itemLink = `zotero://open-pdf/groups/${library.id}/items/${itemKey}`;
      }
      return `<a href="${itemLink}">${_item.getFilename()}</a>`;
    };
    return Zotero.Items.get(topItem.getAttachments())
      .filter((i) => i.isPDFAttachment())
      .map((i) => getPDFLink(i))
      .join("\n");
  })(topItem)}
</p>

## Current Template

```js
// ${await new Promise(async (r) => {
  let header = {};
  header.title = noteItem.parentItem? noteItem.parentItem.getField("title") : "";
  header.date = noteItem.parentItem? 
noteItem.parentItem.getField("date") : "";
  header.alias = noteItem.parentItem? 
noteItem.parentItem.getField("shortTitle") : "";
  header.url = noteItem.parentItem? 
noteItem.parentItem.getField("url") : "";
  header.tags = noteItem.getTags().map((_t) => _t.tag);
  header.collections = (
    await Zotero.Collections.getCollectionsContainingItems([
      (noteItem.parentItem || noteItem).id,
    ])
  ).map((c) => c.name);
  r(JSON.stringify(header));
})}

```

**Comment by windingwind** [comment]
Hi,

> Additionally can you explain to me how you use this in that was in the templates guidance section

To run JS code in Zotero, open menu-> tools -> developers -> run javascript and paste the code there. Make sure a top-level item is selected in the library.

```ts
const item = ZoteroPane.getSelectedItems().shift();
const usedFields = item.getUsedFields();
Zotero.ItemFields.getAll()
.filter((e) => usedFields.indexOf(e.id) >= 0)
.map((e) => e.name);
```

The `noteItem` is the current note you will export/sync. Here the `topItem` equals `noteItem.parentItem` (if it exists). So just replace the `topItem` with `noteItem.parentItem`.

The selection syntax `condition a ? command b : command c` means run `b` if `a` is true (which means, `a` is not empty) and run `c` otherwise. We use `noteItem.parentItem? noteItem.parentItem.getField("title") : ""` to get the title if the note has a parent item and empty string otherwise.

If you still encounter failures, please paste a screenshot of the library with the note selected, which fails with yaml header rendering when exported.

**Comment by Dcrizz** [comment]
Hello

Thanks very much for that!! I will try it out and let you know if I have any problems! Thanks for making this plug-in as well! It has saved me so much time since I started using it!
